janus_collector/src/lib.rs

```diff
@@ -0,0 +1,1032 @@
+    #[error("Timed out waiting for collection to finish")]
+    CollectPollTimeout,
+}
+
+static COLLECTOR_USER_AGENT: &str = concat!(
+    env!("CARGO_PKG_NAME"),
+    "/",
+    env!("CARGO_PKG_VERSION"),
+    "/",
+    "collector"
+);
+
+/// Authentication configuration for communication with the leader aggregator.
+#[derive(Derivative)]
+#[derivative(Debug)]
+#[non_exhaustive]
+pub enum Authentication {
+    /// Bearer token authentication, via the `DAP-Auth-Token` header.
+    DapAuthToken(#[derivative(Debug = "ignore")] AuthenticationToken),
+}
+
+/// The DAP collector's view of task parameters.
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub struct CollectorParameters {
+    /// Unique identifier for the task.
+    task_id: TaskId,
+    /// The base URL of the leader's aggregator API endpoints.
+    #[derivative(Debug(format_with = "std::fmt::Display::fmt"))]
+    leader_endpoint: Url,
+    /// The authentication information needed to communicate with the leader aggregator.
+    authentication: Authentication,
+    /// HPKE configuration and public key used for encryption of aggregate shares.
+    hpke_config: HpkeConfig,
+    /// HPKE private key used to decrypt aggregate shares.
+    #[derivative(Debug = "ignore")]
+    hpke_private_key: HpkePrivateKey,
+    /// Parameters to use when retrying HTTP requests.
+    http_request_retry_parameters: ExponentialBackoff,
+    /// Parameters to use when waiting for a collect job to be processed.
+    collect_poll_wait_parameters: ExponentialBackoff,
+}
+
+impl CollectorParameters {
+    /// Creates a new set of collector task parameters.
+    pub fn new(
+        task_id: TaskId,
+        mut leader_endpoint: Url,
+        authentication_token: AuthenticationToken,
+        hpke_config: HpkeConfig,
+        hpke_private_key: HpkePrivateKey,
+    ) -> CollectorParameters {
+        // Ensure the provided leader endpoint ends with a slash.
+        url_ensure_trailing_slash(&mut leader_endpoint);
+
+        CollectorParameters {
+            task_id,
+            leader_endpoint,
+            authentication: Authentication::DapAuthToken(authentication_token),
+            hpke_config,
+            hpke_private_key,
+            http_request_retry_parameters: http_request_exponential_backoff(),
+            collect_poll_wait_parameters: ExponentialBackoff {
+                initial_interval: StdDuration::from_secs(15),
+                max_interval: StdDuration::from_secs(300),
+                multiplier: 1.5,
+                max_elapsed_time: None,
+                ..Default::default()
+            },
+        }
+    }
+
+    /// Replace the exponential backoff settings used for HTTP requests.
+    pub fn with_http_request_backoff(mut self, backoff: ExponentialBackoff) -> CollectorParameters {
+        self.http_request_retry_parameters = backoff;
+        self
+    }
+
+    /// Replace the exponential backoff settings used while polling for aggregate shares.
```

Comment:
IIUC, this backoff would effectively be overridden by the `Retry-After` value, correct? If so, we should document this.

However, I wonder if these are the desired semantics: currently, if I specify a backoff strategy that says to retry after (say) 10 seconds, and the `Retry-After` header says to retry after 5 seconds, the library will attempt to retry after 5 seconds. Effectively, my `collect_poll_backoff` settings are not being respected in this scenario.

I think better semantics might be to sleep for an amount of time equal to `max(retry-after header, exponential backoff's next wait time)` -- this would respect both of the retry values. WDYT?

(another solution might be to remove backoff configuration entirely, and always rely on `Retry-After`, settling on a reasonable default if the server doesn't set the `Retry-After` header. this is even more opinionated, however.)